crosschain-resolver/contracts/L1Resolver.sol

```diff
@@ -63,12 +66,67 @@ contract L1Resolver is EVMFetchTarget {
     }
 
     /**
-     * Returns the address associated with an ENS node.
-     * @param node The ENS node to query.
-     * @return The associated address.
+     * Query the target address that is verifiable against the name
+     * @param name DNS encoded ENS name to query
+     * @param offset The offset of the label to query recursively. Start from the 0 position and kepp adding the length of each label as it traverse. The function exits when len is 0.
+     * @return node The node of the name
+     * @return target The L2 resolver address to verify against.
+     */
+    function getTarget(
+        bytes memory name,
+        uint256 offset
+    ) public view returns (bytes32 node, address target) {
+        uint256 len = name.readUint8(offset);
+        node = bytes32(0);
+        if (len > 0) {
+            bytes32 label = name.keccak(offset + 1, len);
+            (node, target) = getTarget(
+                name,
+                offset + len + 1
+            );
+            node = keccak256(abi.encodePacked(node, label));
+            if(targets[node] != address(0)){
+              target = targets[node];
+            }
+        } else {
+            return (
+                node,
+                target
+            );
+        }
+        return (node, target);
+    }
+
+    /** 
+     * @dev Resolve and verify a record stored in l2 target address. It supports subname by fetching target recursively to the nearlest parent.
+     * @param name DNS encoded ENS name to query
+     * @param data The actual calldata
+     * @return result result of the call
      */
-    function addr(bytes32 node) public view returns (address) {
-        EVMFetcher.newFetchRequest(verifier, targets[node])
+    function resolve(bytes calldata name, bytes calldata data) external view returns (bytes memory result) {
+        (, address target) = getTarget(name, 0);
+        bytes4 seletor = bytes4(data);
+        if (seletor == 0x3b3b57de) {
```

Comment:
These should be constants - ideally, `SomeInterface.func.selector`.

Reply:
Couldn't figure out how to do that for functions with same name and different arguments (eg: `addr(bytes32,uint)` so used `bytes4(keccak256("addr(bytes32,uint256)"))` instead.